Add setting to control when bridge infill is used

jschuh opened the pull request:
This fixes issue #3814 by adding a bridge_infill_threshold setting to control when bridge infill is applied. The change works by computing the average void area for the infill below the solid layer, and flagging the solid layer as stInternalBridge only if it's larger than the threshold value. Assuming a reasonable threshold is chosen, **this results in both reduced print times and stronger parts with no apparent reduction in print quality.**

The default bridge_infill_threshold value is 25mm², which has been reliable in all my testing with PLA and PETG, and means that on .4mm infill extrusion width you will get bridge infill at 15% infill density but not at 20% infill density.

I've been carrying this change in my local build and printing with it for the better part of a year with good results. It pairs very well with my perimeter fan speed controls in PR #2921, which I rely on when printing any structural parts.

bubnikv commented:
I don't think that this pull request fixes #3814 . Those are two independent topics.

I see some sense in what you have done, however I would rather have the threshold hard coded and pretty low. I suppose for 50% infill the bridging is not really needed.

jschuh commented:
Happy to hardcode it and remove the preference, but how strongly do you feel about that 50% infill threshold? I ask because my feeling is that it's less about the infill percentage and more about the length of the unsupported span between infill extrusions. So, I think it still makes sense to calculate from the extrusion width, which would look something like this:

`unsupported_span = extrusion_width / infill_percentage - extrusion_width
`

Based on my testing 1mm would be a conservative span threshold, which works out to 29% infill percentage for a .4mm extrusion width, and 55% infill percentage for a 1.2mm extrusion width. Accepting that, these are obviously just estimates, because the exact length of the span depends on the infill pattern and the angle of the layer above, but I've found this calculation to be good enough.

jschuh commented:
I removed the setting, now calculate the threshold as described in my last comment, and rolled forward to the latest trunk.

@bubnikv, does this seem reasonable, or would you prefer to skip the extrusion width calculation and just use a very conservative infill density?

jschuh commented:
@bubnikv, I think I've found a very simple and correct solution. The bridge_over_infill() function already had a threshold check near the end to to avoid bridging over voids below a certain width. So, I just moved that threshold calculation up earlier in the code, and also use it to avoid bridging over voids in the infill below the same width.

This is also a much simpler patch, because my earlier version was erroneously performing redundant threshold checks inside the loop, rather than checking before to avoid the loop entirely. And I moved the check entirely inside bridge_over_infill, so no other files or methods are modified.

jschuh commented:
I just squashed and rebased to today's HEAD to make this an easy merge. It's only a 15 line change, localized to a single file, and no longer introduces new configuration options.

I see the current iteration of this patch as effectively a bug fix, because the existing code already skips bridging for open voids less than three times the bridge flow width. So, this patch just extends that same logic to voids in the infill _(e.g. for the typical 0.4mm nozzle with a 0.45mm infill extrusion width this would skip bridging for anything over ~26% infill)_.

I've been printing with this iteration of the patch for the last few months, and I just get faster prints with no loss in quality when printing dense infill (which makes sense since it just applies the existing void threshold).

viper0078 commented:
I usually used a 25% infill. So, I agree with the 26% proposal. My reasons for wanting this feature are described in #7324.
I hope this feature will be implemented soon.
Is it not possible in Ver 2.4?

bubnikv commented:
Sorry it did not make it into 2.4. We need to roll out 2.4. We should look into the pull request after the 2.4 release. 

I agree that for dense infills the current bridging flow over sparse infill is counter productive. For sparse infill, the current behavior is beneficial compared to other slicers, which don't do it.

We need to re-evaluate with both the thick and the new "thin" bridges. The "thin" bridges are what everybody in the industry is doing, the "thick" bridges are what Slic3r always did.

scotfor commented:
> I don't think that this pull request fixes #3814 . Those are two independent topics.
> 
> I see some sense in what you have done, however I would rather have the threshold hard coded and pretty low. I suppose for 50% infill the bridging is not really needed.

I think the threshold should  NOT be hard-coded. It needs to be exposed as a setting.

jschuh commented:
> I think the threshold should NOT be hard-coded. It needs to be exposed as a setting.

The minimum void width that will cause the slicer to use the bridge flow is hardcoded (to 3mm) in the current code. So, it doesn't really make sense to add a separate configuration value for just this one specific case. An earlier version of this PR did use a configuration setting, but in retrospect that was an error on my part, because I didn't understand the code well enough at the time.

scotfor commented:
> > I think the threshold should NOT be hard-coded. It needs to be exposed as a setting.
> 
> The minimum void width that will cause the slicer to use the bridge flow is hardcoded (to 3mm) in the current code. So, it doesn't really make sense to add a separate configuration value for just this one specific case. An earlier version of this PR did use a configuration setting, but in retrospect that was an error on my part, because I didn't understand the code well enough at the time.

Maybe. I would use the threshold to shut bridge infill off completely. Given the number of posts about bridge infill, I think there is little doubt that it is broken the way it is. I just want it gone from my prints. From a tool design perspective it is bad design to have something appear ( bridge infill -  where did that come from?)  that there are no setting for. I can use settings to get solid infill to bridge properly.

jschuh commented:
@scotfor, I'm unclear on what you're proposing, and regardless this PR seems the wrong place for a different proposal. I don't make any decisions for this project (I'm just an independent, open source contributor), and I authored this PR for addressing a specific issue.

My personal suggestion is that you open a new issue specific to the problem you're seeing. I've found the development team to generally be responsive to clear, cogent explanations of problems, along with supporting data. Something that seems obvious in your specific usage could be incomprehensible to someone else—or your understanding of the issue may simply be incorrect. That's why it's so important to be clear in your communication, and respectful to those you're communicating with.

scotfor commented:
> @scotfor, I'm unclear on what you're proposing, and regardless this PR seems the wrong place for a different proposal. I don't make any decisions for this project (I'm just an independent, open source contributor), and I authored this PR for addressing a specific issue.
> 
> My personal suggestion is that you open a new issue specific to the problem you're seeing. I've found the development team to generally be responsive to clear, cogent explanations of problems, along with supporting data. Something that seems obvious in your specific usage could be incomprehensible to someone else—or your understanding of the issue may simply be incorrect. That's why it's so important to be clear in your communication, and respectful to those you're communicating with.

I do have another issue open: https://github.com/prusa3d/PrusaSlicer/issues/7728#issuecomment-1009691308
Lets leave it at that. Thanks for your replies.

jschuh commented:
@Godrak, any chance you'd take a look at this one? It had earlier iterations, but for a while now it just disables bridge flow over sufficiently dense infill (and when the upper extrusions are thick enough).
